pkg/controller/elasticsearch/driver/upgrade_forced.go

```diff
@@ -5,37 +5,68 @@
 package driver
 
 import (
+	"fmt"
+
 	corev1 "k8s.io/api/core/v1"
 
 	"github.com/elastic/cloud-on-k8s/pkg/apis/elasticsearch/v1alpha1"
+	"github.com/elastic/cloud-on-k8s/pkg/controller/elasticsearch/label"
 	"github.com/elastic/cloud-on-k8s/pkg/utils/k8s"
 )
 
 // maybeForceUpgrade may attempt a forced upgrade of all podsToUpgrade if allowed to,
 // in order to unlock situations where the reconciliation may otherwise be stuck
 // (eg. no cluster formed, all nodes have a bad spec).
 func (d *defaultDriver) maybeForceUpgrade(actualPods []corev1.Pod, podsToUpgrade []corev1.Pod) (attempted bool, err error) {
+	actualBySset := podsByStatefulSetName(actualPods)
+	toUpgradeBySset := podsByStatefulSetName(podsToUpgrade)
+
 	attempted = false
-	if len(podsToUpgrade) == 0 {
-		return attempted, nil
-	}
-	if !shouldForceUpgrade(actualPods) {
-		return attempted, nil
-	}
 
-	attempted = true
-	log.Info("Performing a forced rolling upgrade since no Pod is ready",
-		"namespace", d.ES.Namespace, "es_name", d.ES.Name,
-		"pod_count", len(podsToUpgrade),
-	)
-	for _, pod := range podsToUpgrade {
-		if err := deletePod(d.Client, d.ES, pod, d.Expectations); err != nil {
-			return attempted, err
+	for ssetName, actual := range actualBySset {
+		toUpgrade, exists := toUpgradeBySset[ssetName]
+		if !exists || len(toUpgrade) == 0 {
+			continue
 		}
+		if !shouldForceUpgrade(actual) {
+			continue
+		}
+		attempted = true
+		log.Info("Performing a forced rolling upgrade",
+			"namespace", d.ES.Namespace, "es_name", d.ES.Name,
+			"statefulset_name", ssetName,
+			"pod_count", len(podsToUpgrade),
+		)
+		for _, pod := range toUpgrade {
+			if err := deletePod(d.Client, d.ES, pod, d.Expectations); err != nil {
+				return attempted, err
+			}
+		}
+
 	}
+
 	return attempted, nil
 }
 
+func podsByStatefulSetName(pods []corev1.Pod) map[string][]corev1.Pod {
+	byStatefulSet := map[string][]corev1.Pod{}
+	for _, p := range pods {
+		ssetName, exists := p.Labels[label.StatefulSetNameLabelName]
+		if !exists {
+			log.Error(
+				fmt.Errorf("expected label %s not set", label.StatefulSetNameLabelName),
+				"skipping forced upgrade",
+				"namespace", p.Namespace, "pod_name", p.Name)
+			continue
```

Comment:
So if one pod is missing a label we are just dropping that pod from the algorithm and logging an error? I wonder how we could get into such a situation? The only thing I can think of is a user manually deleting a label.

Reply:
Yeah that should definitely not happen :/
Not sure what more I can do here.